### PR TITLE
Documentation: Reorganize the Abstract Models section

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:
This PR reorganizes and cleans up the "Working with Abstract Models" section.  In particular, it:
 - adds information on using `create_instance`
 - moves around some of the data initialization information to get it all together
 - removes redundant (nearly repeated) text
...plus a few other minor edits that sneaked in (see the changes below)

## Changes proposed in this PR:
- Reorganize Abstract Models section of the documentation
- Adds Python 3.7 as a supported version
- Adds autodoc generation for more core classes (with lots of ensuing warnings, to be fixed later)
- removed use of deprecated `cname()` in examples
- cleans up class references (use the shorter "``:class:`~pyomo.environ.Set` `` notation for referencing `Set`

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by DLWoodruff** [comment]
This looks very good to me, except I don't like the deletion of some lines that seem to me to important for novice users (who often complain: "I typed in just what was in the documentation and it didn't work"). These lines were intended to give them a hint and to give their instructors some help:

Instantiating Models
-------------------- 
If you have a `ConcreteModel` it will be constructed as components are
declared. If you have an `AbstractModel` you must instantiate it to
create a concrete instance. 
.. note:: 
   `AbstractModel` users should note that your concrete model instance
   is sometimes called "instance" and **not** "model". This is the case
   when you use `instance = model.create_instance()`.

**Comment by jsiirola** [comment]
@DLWoodruff: Oops.  There was a whole file missing from the original PR.  See the (now included) "instantiating_models.rst".  I also added your `.. note::` back into that file as well (to reinforce the point).

File: doc/OnlineDocs/working_abstractmodels/data/dataportals.rst [hunk 1]
```diff
@@ -143,8 +139,54 @@ loaded:
...
+
+The data in this file can be used to load the following model:
+
+.. literalinclude:: ../../tests/dataportal/dataportal_tab_json1.spy
+    :language: python
+
+Note that no ``set`` or ``param`` option needs to be specified when
+loading a ``JSON`` or ``YAML`` file.  All of the set and parameter
+data in the file are loaded by the :class:`DataPortal <ref
+pyomo.core.base.DataPortal.DataPortal>` object, and only the data
```
Comment (blnicho) [inline doc/OnlineDocs/working_abstractmodels/data/dataportals.rst]: Looks like you changed all of the other `:class:` references in this section to have a slightly different format, did you miss this one?

File: doc/OnlineDocs/working_abstractmodels/data/index.rst [hunk 2]
```diff
@@ -1,51 +1,50 @@
...
+===============================
 
 There are roughly three ways of using data to construct a Pyomo
 model:
 
-1. use standard Python data objects, 
+1. use standard Python objects,
 
-2. initialize a model with data loaded with a :class:`DataPortal <ref pyomo.core.base.DataPortal.DataPortal>` object, and
+2. initialize a model with data loaded with a :class:`DataPortal <ref pyomo.environ.DataPortal>` object, and
```
Comment (blnicho) [inline doc/OnlineDocs/working_abstractmodels/data/index.rst]: Same comment as before, did you mean to change the formatting of this class reference?

**Comment by blnicho** [comment]
This is only changing documentation and the updated doctests were shown to pass in the second to last commit so I'm going to merge this without waiting for the tests to finish.

**Pull request merged**